compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp

```diff
@@ -113,6 +113,163 @@ struct FoldAttentionMaskUnitDim final
+/// ```
+/// %0 = expand_shape ... tensor<3x3x10xf32> into tensor<3x3x5x2xf32>
+/// %1 = collapse_shape %1 ... tensor<3x3x5x2xf32> into tensor<9x5x2xf32>
+/// ```
+struct DropUnitDimsFromCollapseOfExpand
+    : OpRewritePattern<tensor::CollapseShapeOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::CollapseShapeOp collapseOp,
+                                PatternRewriter &rewriter) const override {
+    auto expandOp = collapseOp.getSrc().getDefiningOp<tensor::ExpandShapeOp>();
+    if (!expandOp) {
+      return failure();
+    }
+
+    const SmallVector<ReassociationIndices, 4> collapseReassoc =
+        collapseOp.getReassociationIndices();
+    ArrayRef<int64_t> interShape = expandOp.getType().getShape();
+    ArrayRef<int64_t> outShape = collapseOp.getType().getShape();
+    SmallVector<int64_t> interToOutMap(expandOp.getType().getRank());
+
+    // Step 1: construct a set of dimensions (with respect to the intermediate
+    // shape) that are unit length and get collapsed away by the final
+    // `collapse_shape` op.
+    llvm::SmallDenseSet<int64_t> toDrop;
+    for (const auto &[outDim, indices] : llvm::enumerate(collapseReassoc)) {
+      for (auto [innerIdx, inDim] : llvm::enumerate(indices)) {
+        // Can't drop this dim if it isnt statically 1 or if it isn't being
+        // combined with any other dimensions.
+        if (indices.size() == 1 || interShape[inDim] != 1) {
+          continue;
+        }
+
+        // If we are collapsing multiple unit dims together, at least 1 must be
+        // kept (prefer the first).
+        if (outShape[outDim] == 1 && innerIdx != 0) {
+          continue;
+        }
+        toDrop.insert(inDim);
+      }
+    }
+
+    // Step 2: remove dimensions from `toDrop` that weren't introduced by the
+    // `expand_shape` op.
+    const auto expandReassoc = expandOp.getReassociationIndices();
+    for (const auto &[inDim, indices] : llvm::enumerate(expandReassoc)) {
+      if (indices.size() == 1) {
+        toDrop.erase(indices[0]);
+      }
+    }
+    if (toDrop.empty()) {
+      return rewriter.notifyMatchFailure(collapseOp,
+                                         "Didn't find any unit dims to drop");
+    }
+
+    // Step 3: construct a new intermediate shape without the foldable dims.
+    SmallVector<int64_t> newInterShape;
+    SmallVector<OpFoldResult> newOutputSizes;
+    newInterShape.reserve(interShape.size() - toDrop.size());
+    newOutputSizes.reserve(interShape.size() - toDrop.size());
+    SmallVector<OpFoldResult> outputSizes = getMixedValues(
+        expandOp.getStaticOutputShape(), expandOp.getOutputShape(), rewriter);
+    for (auto [idx, ofr] : llvm::enumerate(outputSizes)) {
+      if (!toDrop.contains(idx)) {
+        std::optional<int64_t> staticDim = getConstantIntValue(ofr);
+        newInterShape.push_back(staticDim.value_or(ShapedType::kDynamic));
+        newOutputSizes.push_back(ofr);
+      }
+    }
```

Comment:
Output is a little ambiguous to me because it could be either the output of the expand shape op or collapse shape op. Maybe rename it to `origInterSizes` and `newInterSizes`?